### ISSUE TITLE
Cache non-deterministic transforms for SmartCacheDataset

### ISSUE THREAD

**Issue by EdvardGrodem** [Project-MONAI/MONAI#4059]
To increase speed it would be nice to have the option to cache non-deterministic transforms when using [SmartCacheDataset](non-deterministic). In this way, only the new items will be processed, which will significantly increase the transform speed. 

For instance, let's say replace_rate=0.25 is used. Then every item in the dataset will use the same non-deterministic transforms 4 times, and the increase in speed is potentially 4 times faster. I believe that even though the same non-deterministic transforms will be used in 4 epochs following each other, this will not have a signific influence on overfitting since the next time the same data is loaded a new non-deterministic will be applied.

**Comment by Nic-Ma** [comment]
Hi @EdvardGrodem ,

Thanks for your feedback and interesting idea.
Seems your idea can accelerate the training time of every epoch, but I am not sure whether the overall training time is decreased when achieving the **target metrics**.
If we decrease the random augmentations, it may require more epochs to train, better to have some experiments first.
@ericspod @dongyang0122 @wyli Do you have any comments from research practice?

Thanks in advance.

**Comment by EdvardGrodem** [comment]
Thanks for your answers. I think you have an interesting point @dongyang0122. A better solution would maybe be to have the option to give two transforms to the SmartCacheDataset. One is performed when caching the data and another before loading the data to a batch. In this way, heavy augmentations such as affine transformations can be performed once, and then lightweight augmentations such as cropping or masking can be done when loading a batch. 

For the model I'm working on it would still be beneficial to just cache all the random transforms, since random masking is done by the model itself for self-supervised training.

**Comment by vikashg** [comment]
closing because of inactivity
